File: 189-rotate-array/rotate-array.kt
class Solution {
    fun rotate(nums: IntArray, k: Int) {
        val n = nums.size
        val rotations = k % n // Handle cases where k > n

        if (rotations == 0) return // No rotation needed

        val temp = nums.copyOfRange(n - rotations, n) // Store the last 'rotations' elements
        for (i in n - 1 downTo rotations) {
            nums[i] = nums[i - rotations] // Shift the first (n - rotations) elements to the right
        }
        for (i in 0 until rotations) {
            nums[i] = temp[i] // Place the stored last 'rotations' elements at the beginning
        }
    }
}

File: 189-rotate-array/rotate-array.py
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        k = k %len(nums)
        n = len(nums)
        nums[0:0] = nums[n-k:]
        nums[n:] = []
